crates/core/src/parser/parser.rs

```diff
@@ -1140,6 +1147,48 @@ mod test {
         assert_eq!(parse_variable(variable), expected);
     }
 
+    #[test]
+    fn test_parse_variable_with_dot_separated_array_index() {
+        let variable = LiquidParser::parse(Rule::Variable, "foo.0.bar.baz[foo.bar]")
+            .unwrap()
+            .next()
+            .unwrap();
+
+        let indexes = vec![
+            Expression::Literal(Value::scalar(0)),
+            Expression::Literal(Value::scalar("bar")),
+            Expression::Literal(Value::scalar("baz")),
+            Expression::Variable(Variable::with_literal("foo").push_literal("bar")),
+        ];
+
+        let mut expected = Variable::with_literal("foo");
+        expected.extend(indexes);
+
+        assert_eq!(parse_variable(variable), expected);
+    }
+
+    #[test]
+    fn test_parse_variable_with_dot_separated_array_index2() {
+        let variable = LiquidParser::parse(Rule::Variable, "foo.0.bar.2.4.baz[foo.bar]")
```

Comment:
will this case also work and is it valid?
```
foo.0.bar.-2.+4
```

Reply:
yep
 A negative index will count from the end of the array.
 +4 = 4